# Scripts for
## Wang, X., ... Heller, R. (2025). Population structure and domestication history of the Javan banteng.

### evaluation of assembly


### Mapping
A pipeline for mapping and post-mapping filtering using PALEOMIX BAM pipeline.

### Sample filtering
Script to assess post-mortem damage in historical samples using mapDamage2;

A pipeline designed to remove problematic samples prior to downstream analyses;

Script to perform a second and more sensitive analysis of relatedness within each sampling location using NgsRelate.

### Sites filtering
A pipeline designed to avoid biases from low-quality mapping sites.

### Genotype likelihood calculation
A pipeline designed to call genotype likelihood for output as BEAGLE.

### Genotype calling, imputation and phasing
A pipeline designed to call genotypes;

Scripts to imputate and phase genotypes.

### Population structure and admixture analyses
Script to infer principal component analysis (PCA) using HaploNet;

Script to estimate admixture proportions for each individual using NGSadmix;

Script to use evalAdmix;

Script to perform apoh.

### Genetic diversity, divergence and inbreeding
Script to calculate genome-wide heterozygosity;

Script to infer Pairwise global Fst between each pair of populations;

Runs of homozygosity: 

Script to infer runs of homozygosity (ROH);

Script to calculate genome-wide heterozygosity without ROH.

### Population history and introgression
Script to infer the population tree assuming different numbers of admixture events using TreeMix;

Script to infer the ancient admixture events, we calculated D statistics (ABBA-BABA) using the ANGSD;

Script to infer the f3-outgroup using the R package ADMIXTOOLS2.

### Demographic history
Script to infer effective population size changes using PSMC;

Script to estimate domestication event between wild (Banteng-WildBorn) and domesticated (Bali-Bali) using Fastsimcoal2;

Script to estimate Confidence interval using Fastsimcoal2;

All 14 demographic model design for estimating domestication event by Fastsimcoal2.

### Genome-wide selection scans
Script to perform sliding window Fst, nucleotide diversity and Tajimaâ€™s D;

Script to infer linkage disequilibrium by ngsLD;

Script to infer XP-EHH.

### Genetic load
Scripts to infer genetic load.
