# Scripts for (still working on)
## Wang, X., ... Heller, R. (2025). Population structure and domestication history of the Javan banteng (Bos javanicus javanicus)

### Mapping
A pipeline for mapping and post-mapping filtering using PALEOMIX BAM pipeline.

### Sample filtering
A pipeline designed to remove problematic samples prior to downstream analyses.

### Sites filtering
A pipeline designed to avoid biases from low-quality mapping sites.

### Genotype likelihood calculation
A pipeline designed to call genotype likelihood for output as BEAGLE;

### Genotype calling
A pipeline designed to call genotypes;

### Imputation and phasing
Scripts to imputate and phase genotypes;

### Population structure
Script to infer principal component analysis (PCA) using HaploNet;

Script to estimate admixture proportions for each individual using NGSadmix;

Script to use evalAdmix;

### Genetic diversity, divergence and inbreeding
Genome-Wide heterozygosity: Script to calculate genome-wide heterozygosity;Script to infer the ancient admixture events, we calculated D statistics (ABBA-BABA) using the R package ADMIXTOOLS2;

Pairwise global Fst: Script to infer Pairwise global Fst between each pair of populations;

Runs of homozygosity: 

Script to infer runs of homozygosity (ROH);

Script to calculate genome-wide heterozygosity without ROH;

### Population history and introgression
Treemix: Script to infer the population tree assuming different numbers of admixture events using TreeMix;

Dstatistics: Script to infer the ancient admixture events, we calculated D statistics (ABBA-BABA) using the ANGSD;

f3-Outgroup: Script to infer the f3-outgroup using the R package ADMIXTOOLS2;

### Demographic history
Script to estimate domestication event between wild (Banteng-WildBorn) and domesticated (Bali-Bali) using Fastsimcoal27;

Script to estimate Confidence interval using Fastsimcoal27;

All 14 demographic model design for estimating domestication event by Fastsimcoal27;

Script to infer effective population size changes using PSMC;

### Genome-wide selection scans
Script to perform sliding window Fst, nucleotide diversity and Tajimaâ€™s D;

Script to infer linkage disequilibrium by ngsLD;

Script to infer XP-EHH;

### Genetic load


